compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/VectorizeIREEVectorExtOps.cpp

```diff
@@ -121,4 +125,365 @@ struct VectorizeIREEVectorExtOpsPass final
+        map = map.insertResult(getAffineDimExpr(indexOp.getDim(), ctx),
+                               map.getNumResults());
+        found = true;
+        break;
+      }
+
+      // If this index is a value defined outside the linalg.generic block, it
+      // is a constant value in the loop.
+      Operation *defOp = trace.getDefiningOp();
+      if (defOp && !genericOp.getBody()->findAncestorOpInBlock(*defOp)) {
+        map = map.insertResult(getAffineConstantExpr(0, ctx),
+                               map.getNumResults());
+        found = true;
+        break;
+      }
+
+      // If this index is a block argument, the index is a loop iterated
+      // variable, and the iteration space of the index is same as the loop
+      // iterated variable. However, we restrict to cases where the variable
+      // iteration space is 1-D or 0-D, otherwise the Memory -> Loop Iteration
+      // Space map will contain mods/floordivS.
+      if (auto blockArg = dyn_cast<BlockArgument>(trace)) {
+        // If this block argument isn't part of the linalg.generic body, then
+        // this is a constant value in the loop.
+        if (blockArg.getParentBlock() != genericOp.getBody()) {
+          map = map.insertResult(getAffineConstantExpr(0, ctx),
+                                 map.getNumResults());
+          found = true;
+        } else {
+          // Allow 0-D and 1-D maps.
+          AffineMap indexMap =
+              genericOp.getIndexingMapsArray()[blockArg.getArgNumber()];
+          if (indexMap.getNumResults() == 0) {
+            map = map.insertResult(getAffineConstantExpr(0, ctx),
+                                   map.getNumResults());
+            found = true;
+          } else if (indexMap.getNumResults() == 1) {
+            if (auto dimExpr = dyn_cast<AffineDimExpr>(indexMap.getResult(0))) {
+              map =
+                  map.insertResult(getAffineDimExpr(dimExpr.getPosition(), ctx),
+                                   map.getNumResults());
+              found = true;
+            }
+          }
+        }
+        break;
+      }
+
+      // Trace back arith.index_cast ops. This is a commonly occuring case,
+      // there may be other cases which we can trace back here (For example,
+      // addition with a loop invariant constant).
+      if (isa<arith::IndexCastOp>(defOp)) {
+        trace = defOp->getOperand(0);
+      } else {
+        break;
+      }
+    }
+
+    if (!found) {
+      return AffineMap();
+    }
+  }
+
+  return map;
+}
+
+static AffineMap inversePerm(AffineMap map) {
+  MLIRContext *context = map.getContext();
+  AffineExpr zero = mlir::getAffineConstantExpr(0, context);
+  // Start with all the results as 0.
+  SmallVector<AffineExpr, 4> exprs(map.getNumInputs(), zero);
+  for (unsigned i : llvm::seq(unsigned(0), map.getNumResults())) {
+    // Skip zeros from input map. 'exprs' is already initialized to zero.
+    if (auto constExpr = dyn_cast<AffineConstantExpr>(map.getResult(i))) {
```

Comment:
You can do `map.isProjectedPermutation(/*allowZeroInResults=*/true)` to skip the need to check the values of the constants here.

Reply:
The check is to allow zero constants in the results and skip over them. I think i'll need to still do the check.